### PR TITLE
Allow null DropdownButton values

### LINKED ISSUE #6422: Dropdown null selection & hints
Feature request from Customer Leafy:

- [ ] Add the ability to show a dropdown which does not have a selection (yet)
- [ ] Add "hint" text to display in the dropdown when no value is selected
- [ ] Add ability to show error-text for dropdowns, as we'd need to show that a "None" selection would not be valid when submitting a form

### PR DESCRIPTION
**Pull request by HansMuller**
Specifying null as the DropdownButton's value causes the dropdown to appear empty until a selection is made. 

If the dropdown is added to an InputContainer, a hint and label can be displayed as well.

Changed IndexedStack: specifying index: null prevents the widget from displaying anything or responding to input.

Fixes #6422

Here's an example that demos this feature along with InputContainer.

```
import 'package:flutter/material.dart';

class InputContainerDemo extends StatefulWidget {
  @override
  _InputContainerDemoState createState() => new _InputContainerDemoState();
}

class _InputContainerDemoState extends State<InputContainerDemo> {
  static final List<String> _dropdownItems = <String>['One', 'Two', 'Free', 'Four'];
  String _dropdownValue = null;

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(title: new Text('InputContainer Demo')),
      body: new Padding(
        padding: const EdgeInsets.all(32.0),
        child: new DropdownButtonHideUnderline(
          child: new Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: <Widget>[
              new InputContainer(
                labelText: _dropdownValue == null ? 'Where are you from' : 'From',
                isEmpty: _dropdownValue == null,
                child: new DropdownButton<String>(
                  value: _dropdownValue,
                  isDense: true,
                  onChanged: (String newValue) {
                    setState(() {
                      _dropdownValue = newValue;
                    });
                  },
                  items: _dropdownItems.map((String value) {
                    return new DropdownMenuItem<String>(
                      value: value,
                      child: new Text(value),
                    );
                  }).toList(),
                ),
              ),
              new SizedBox(height: 32.0),
              new Center(
                child: new RaisedButton(
                  onPressed: () {
                    setState((){
                      _dropdownValue = null;
                    });
                  },
                  child: new Text('CLEAR'),
                ),
              ),
            ],
          ),
        ),
      )
    );
  }
}

void main() {
  runApp(new MaterialApp(home: new InputContainerDemo()));
}
```

### CODE REVIEW

File: packages/flutter/lib/src/material/drop_down.dart [hunk 1]
```diff
@@ -533,10 +538,15 @@ class _DropdownButtonState<T> extends State<DropdownButton<T>> {
           mainAxisAlignment: MainAxisAlignment.spaceBetween,
           mainAxisSize: MainAxisSize.min,
           children: <Widget>[
-            new IndexedStack(
-              index: _selectedIndex,
-              alignment: FractionalOffset.centerLeft,
-              children: config.items
+            // The button's size is defined by its largest menu item. If value is
+            // null then an item does not appear.
+            new Opacity(
```
Comment (abarth) [inline packages/flutter/lib/src/material/drop_down.dart]: Why use an opacity here?  I guess that works, but I would have expected there to be a more direct way of making a subtree take up space but be invisible.

In particular, the children can still be hit with opacity 0.0, but presumably that's not what you want.

File: packages/flutter/lib/src/material/drop_down.dart [hunk 2]
```diff
@@ -533,10 +538,15 @@ class _DropdownButtonState<T> extends State<DropdownButton<T>> {
...
+              opacity: _selectedIndex == null ? 0.0 : 1.0,
+              child: new IndexedStack(
+                index: _selectedIndex ?? 0,
```
Comment (abarth) [inline packages/flutter/lib/src/material/drop_down.dart]: Maybe you should teach indexedstack to take a null index, which means "show none of the children"

Comment (HansMuller) [inline packages/flutter/lib/src/material/drop_down.dart]: Sure, sounds good, I'll do that.

**Comment by abarth** [comment]
LGTM

File: packages/flutter/lib/src/material/drop_down.dart [hunk 3]
```diff
@@ -411,28 +412,31 @@ class DropdownButtonHideUnderline extends InheritedWidget {
 class DropdownButton<T> extends StatefulWidget {
```
Comment (Hixie) [inline packages/flutter/lib/src/material/drop_down.dart]: at some point we should document `T` here and in the related classes (not necessarily in this PR unless you're feeling particularly heroic)

Comment (HansMuller) [inline packages/flutter/lib/src/material/drop_down.dart]: I will put it on my TODO list.

https://github.com/flutter/flutter/issues/6975

**Pull request merged**